EKF: cope better with GPS jamming

tridge opened the pull request:
This is a revival of https://github.com/ArduPilot/ardupilot/pull/23010
re-opened so I can push to it to get it to pass CI

Commit:
AP_NavEKF3: Apply GPS quality checks following loss of 3D fix if velocity error is bounded

Commit:
Tools: re-work copter and plane loss of GPS auto tests

Explicitly test time taken to reset to GPS loss and regain of lock for copter without and plane with dead reckoning assistance.

Commit:
AP_NavEKF3: Use last observed wind states to enable dead reckoning

These changes enable the EKF to use the last observable wind velocity vector estimate to synthesise an airspeed measurement if operating without an airspeed sensor and when all other measurement types that can constrain velocoty drift are lost. This enables the EKF to use dead reckoning to continue after loss of GPS when there is no air speed sensor fitted and without the need to set a default airspeed value.
The logic used to fuse a default airspeed value has also been cleaned up and the call to FuseAirSpeed() from inside SelectBetaDragFusion() has been removed.

AP_NavEKF3: Fix error in default airspeed observation variance

AP_NavEKF3: Enable shadow fusion of airspeed when sensor is disabled

Commit:
AP_NavEKF3: Strengthen protection against GPS jamming

These changes prevent the EKF from consuming GPS data too soon when it is recovering from jamming if the EKF is able to navigate using dead reckoning.

magicrub commented on libraries/AP_AHRS/AP_AHRS.cpp:
```diff
@@ -178,6 +178,13 @@ const AP_Param::GroupInfo AP_AHRS::var_info[] = {
 
+    // @Param: OPTIONS
+    // @DisplayName: Optional AHRS behaviour
+    // @Description: This conrols optional AHRS behaviour. Setting DisableDCMFallback will change the AHRS behaviour for fixed wing aircraft to not fall back to DCM when the EKF stops fusing GPS measurements while there is 3D GPS lock
```
Spelling

rmackay9 replied:
I've got a minor commit to fix this.

tridge commented:
rebaed

rmackay9 commented:
I think we should try and get this in the near-ish future.  Certainly in enough time that it gets some testing before 4.5.0 beta testing.

tridge commented:
@priseborough this is failing the dead-reckoning CI test. The EK3 position is resetting on loss of GPS for some reason.
I've put a replay log here:
http://uav.tridgell.net/tmp/ArduPlane-DeadreckoningNoAirSpeed-00000003.BIN

tridge commented:
![image](https://github.com/ArduPilot/ardupilot/assets/831867/d1d44425-f1e8-4191-b9f5-a498630eef2a)

priseborough commented:
> @priseborough this is failing the dead-reckoning CI test. The EK3 position is resetting on loss of GPS for some reason.
> I've put a replay log here:
> http://uav.tridgell.net/tmp/ArduPlane-DeadreckoningNoAirSpeed-00000003.BIN

The underlying issue is that the EKF in the no airspeed sensor case is not fusing in a synthetic airspeed to constrain along track drift and is falling out of dead reckoning within a few seconds of GPS being lost. Setting FLIGHT_OPTIONS bit 7 will cause the EKF to use a default airspeed halfway between ARSPD_FBW_MAX and ARSPD_FBW_MIN which for SITL is between 30 and 10. I will see if I can adjust the SITL parameters so that the vehicle cruises at 20 m/s with no airspeed sensor and set that bi in FLIGHT_OPTIONS and get the dead reckoning working for the no airspeed sensor case.

priseborough committed:
AP_NavEKF3: Fix bug preventing use of default or synthetic airspeed

priseborough committed:
AP_NavEKF3: Treat wind as truth when deadreckoning with no airspeed sensor

priseborough committed:
Tools: Allow dead reckoning test longer to learn wind if no aspd sensor

priseborough committed:
AP_NavEKF3: Allow wind to relearn rapidly when GPS is re-enabled

priseborough committed:
AP_NavEKF3: Fix bug causing in flight yaw align to not complete

priseborough committed:
AP_NavEKF3: Don't block no compass planes from running GPS alignment checks

priseborough commented:
There are multiple contributing  issues:

1. A bug that prevented synthetic or default airspeed being used for dead reckoning  if airspeed data was being received but had been disabled by parameter.
2. Synthetic airspeed and sideslip was not constraining velocity drift sufficiently. For the no airspeed sensor use case using wind states learned prior to loss of GPs as a truth  'truth' reference has fixes this.
3. Learning wind takes longer if there is no airspeed sensor and the autotest was not allowing sufficient time.
4. The GPS jamming was not causing the EKF to stop using the GPS completely, instead because the duration of loss of lock and large errors was relatively short the EKF was instead using the innovation consistency checks to protect against bad estimates. This is actually working, bit not what the autotest was looking for which was a delayed resumption of GPs use while quality checks waited to pass.

There were other issues found:

1. Planes might never register the in flight yaw alignment complete if the yaw misalignment was small
2. When resuming GPS use wind state learning should be faster

I am currently testing fixes for these and will submit a PR against the branch.

priseborough committed:
AP_NavEKF3: Log gpsGoodToAlign

priseborough committed:
AP_NavEKF3: Rework GPS jamming resiliency

Make it user selectable.
Remove potential for a race condition between decisions based on latest data and the EKF fusion processing which operates on a delayed time horizon. This is achieved by preventing data entering the buffer if awaiting checks to pass ensuring that no EKF fusion time horizon processes can use data that hasn't passed checks.
Log the waitingForGpsChecks class variable

priseborough committed:
Tools: Use GPS jamming option in EKF dead reckoning autotests

priseborough committed:
AP_NavEKF3: Rework method of synthesising airspeed for dead reckoning

The previous method resulted in data incest and fusion of predicted airspeed on every EKF internal time step.
This was not apparent during flight where the vehicle was turning, but during long straight legs did not constrain along track drift.

priseborough committed:
Tools: Disable DCM fallback for plane dead reckoning tests

priseborough committed:
Tools: relax req accuracy for plane dead reckoning when not using airspeed

priseborough committed:
Tools: update AHRS_OPTIONS for dead reckoning test

tridge commented:
rebased

tridge commented:
rebased and cleaned up AP_AHRS options handling code

tridge commented:
@priseborough the Deadreckoning test is failing in an interesting way:
![image](https://github.com/ArduPilot/ardupilot/assets/831867/77749761-7b94-4c99-9180-542814aa504e)

tridge committed:
autotest: enable LOG_REPLAY in deadreckoning test

rmackay9 commented:
No pressure but I wonder if we should try and include this in 4.5.0.

rmackay9 commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -734,6 +734,13 @@ const AP_Param::GroupInfo NavEKF3::var_info2[] = {
     AP_GROUPINFO("GPS_VACC_MAX", 10, NavEKF3, _gpsVAccThreshold, 0.0f),
 
+    // @Param: OPTIONS
+    // @DisplayName: Optional EKF behaviour
```
Normally the Display name should start with the feature or subsystem so a better name would be, "EKF3 Optional Behaviour".

For the Description we should remove the first part, "This controls optional EKF behaviour.".  The "This controls" part in particular is not necessary because all parameters control something.

rmackay9 replied:
I've created a commit at the end of this branch to resolve this and fix two other formatting issues. https://github.com/rmackay9/rmackay9-ardupilot/commits/tridge-pr-ekf-gps-check-fix/

priseborough commented:
I'll pick this up again this week

Commit:
autotest: Reduce time threshold used in plane deadreckoning test

priseborough commented:
@tridge, that glitch in height is due to the autotester disarming the plane after the resumption of jamming minimum time test fails. I have determined from dataflash logs that the code is resuming fusion after the gpsGoodToAlign flag goes high, eg:

![GPS_Accuracy](https://github.com/ArduPilot/ardupilot/assets/3596952/a788ef05-5b0b-411f-bd38-070550c25f32)
![Velocity_Innovation_NE](https://github.com/ArduPilot/ardupilot/assets/3596952/a9acfad3-4925-4733-9a8c-81e4a3095c69)
![XKFS 0 GPS_GTA](https://github.com/ArduPilot/ardupilot/assets/3596952/02acb2e7-a50b-42d7-a7f8-6a818b47c066)

I've amended the time threshold which allows the dead reckoning test to pass - https://github.com/priseborough/ardupilot/commit/ae2e1bff9ae9fdf2847e41750a02380df9ac2037

rmackay9 commented:
@priseborough @tridge, great to see progress on this.

Sorry to be nitpicky but could someone pull in the last commit from this library to fixup the new OPTIONS parameter description?
https://github.com/rmackay9/rmackay9-ardupilot/commits/tridge-pr-ekf-gps-check-fix/

tridge commented:
> I've amended the time threshold which allows the dead reckoning test to pass - [priseborough@ae2e1bf](https://github.com/priseborough/ardupilot/commit/ae2e1bff9ae9fdf2847e41750a02380df9ac2037)

thanks Paul! I've added that commit to this branch and rebased on master

rmackay9 reviewed (changes requested):
Could we pull in this change to the new EK3_OPTIONS parameter descriptions?  I'm sorry to be so persistent. https://github.com/rmackay9/rmackay9-ardupilot/commits/tridge-pr-ekf-gps-check-fix/

tridge commented:
> Could we pull in this change to the new EK3_OPTIONS parameter descriptions? I'm sorry to be so persistent. https://github.com/rmackay9/rmackay9-ardupilot/commits/tridge-pr-ekf-gps-check-fix/

applied, sorry for being slow on this!

rmackay9 commented:
yeah!! great to see this go in!

rmackay9 commented:
Just for accountability purposes so that we can all do better next time, this PR broke [Copter deadreckoning](https://ardupilot.org/copter/docs/deadreckoning-failsafe.html).  Here's the issue: https://github.com/ArduPilot/ardupilot/issues/33451 and PR: https://github.com/ArduPilot/ardupilot/pull/33607
